### PR TITLE
Use selection buffer in ray queries (ogre2)

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

* requires https://github.com/ignitionrobotics/ign-gazebo/pull/969

Replaces part of ray query logic with the use of selection buffer (mouse picking) for performance improvement.

The current ray query implementation has 2 stages:
1. ray-bounding box check - broad phase check to narrow down the number of objects that are potentially clicked by the mouse
2. ray-triangle intersection test - checks against all objects found in 1 and compute the actual 3d intersection point on the mesh.

These 2 stages are all done on the CPU and they can be slow if the user clicks on a region with multiple overlapping bounding boxes. This PR replaces stage 1 with the use of selection buffer which can efficiently return the first object that the user clicks on. So that in stage 2 we would only need to do ray intersection test with 1 mesh.

Stage 2 can also be improved. That'll come later. **Update**: see #383

## Test it

Use ign-gazebo, insert multiple overlapping complex meshes, and try panning with the mouse by left clicking on a mesh in the scene and dragging the mouse. The movement should be less laggy with these changes. See https://github.com/ignitionrobotics/ign-gazebo/pull/969

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: ogre2/src/Ogre2RayQuery.cc [hunk 1]
```diff
@@ -25,12 +25,19 @@
...
+#include "ignition/rendering/ogre2/Ogre2SelectionBuffer.hh"
 
 /// \brief Private data class for Ogre2RayQuery
 class ignition::rendering::Ogre2RayQueryPrivate
 {
   /// \brief Ogre ray scene query object for computing intersection.
   public: Ogre::RaySceneQuery *rayQuery = nullptr;
+
+  //// \brief Pointer to camera
+  public: Ogre2CameraPtr camera;
```
Comment (ahcorde) [inline ogre2/src/Ogre2RayQuery.cc:37]:
```suggestion
  public: Ogre2CameraPtr camera{nullptr};
```

File: ogre2/src/Ogre2RayQuery.cc [hunk 2]
```diff
@@ -25,12 +25,19 @@
...
+
+  /// Image pos to cast the ray from
+  public: math::Vector2i imgPos;
```
Comment (ahcorde) [inline ogre2/src/Ogre2RayQuery.cc:40]:
```suggestion
  public: math::Vector2i imgPos = math::Vector2i::Zero;
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ogre2/src/Ogre2RayQuery.cc (hunk 2 above)
Comment (iche033) [reply ogre2/src/Ogre2RayQuery.cc]: 50aefa72c1ed556b718bedd9f88491c2aea55cf1

File: ogre2/src/Ogre2RayQuery.cc (hunk 1 above)
Comment (iche033) [reply ogre2/src/Ogre2RayQuery.cc]: 50aefa72c1ed556b718bedd9f88491c2aea55cf1

**Comment by iche033** [comment]
the `UNIT_Utils_TEST` failure is fixed in part 2, #383

**Review by nkoenig — APPROVED** [review]

File: ogre2/src/media/materials/scripts/selection_buffer.material [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+material SelectionBuffer
+{
+  // Material has one technique
+  technique
+  {
+    // This technique has one pass
+    pass
+    {
+//      fog_override true
```
Comment (ahcorde) [inline ogre2/src/media/materials/scripts/selection_buffer.material:19]: remove ?

**Review by ahcorde — APPROVED** [review]
just a minor comment

File: ogre2/src/media/materials/scripts/selection_buffer.material (hunk 3 above)
Comment (iche033) [reply ogre2/src/media/materials/scripts/selection_buffer.material]: removed. 72bea5e

**Pull request merged**